# Titanic: Machine Learning From Disaster
### by Sung Ahn and Abdul Saleh

<hr>
## Introduction
In this project, we use random forests and (gradient boosting) machine learning algorithms to predict who survived the sinking of the RMS Titanic. On our journey to achieving this goal, we go through the whole data science process from understanding the problem and getting the data to fine-tuning our models and visualizing our results. 

The Titanic dataset is perhaps the most widely analyzed dataset of all time. There exists a wealth of incredible tutorials online exploring different approaches to analyzing this dataset. So in our own analysis, we draw on the experiences of the huge community of amazing people who have already attempted this problem and shared their conclusions online. We would especially like to thank [Manav Sehgal](https://www.kaggle.com/startupsci/titanic-data-science-solutions), [Jeff Delaney](https://www.kaggle.com/jeffd23/scikit-learn-ml-from-start-to-finish), and [Ahmed Besbes](https://ahmedbesbes.com/how-to-score-08134-in-titanic-kaggle-challenge.html) whom without their insight this project would not have become a reality.    

<hr>
## Outline 
1. Understanding the problem
2. Getting the data
3. Exploring the data 
4. Picking a machine learning algorithm 
5. Preparing the data for machine learning algorithms
6. Training algorithm and fine-tuning model
7. Visualizing results and presenting solution 


<hr>
## Understanding the problem
Before we dive into the data analysis and algorithms, we first ask ourselves: is this even a problem that can be solved with machine learning? <br>
Luckily for us, lots of books have been written about the sinking of the Titanic. So before we look at the data, we do some background reading and discover that some patterns might exist, most notably: 

1. Women and children generally got first priority on the life boats 
    - This tells us that we should look for 
2. 
    -
3. There was a lot of confusion during the sinking of the ship and people chose to stay on the Titanic for arbitrary reasons. 
    - There are definitley anomalies in this dataset because it is clear that some people 
    through we conclused th

Aha, so it seems like a there are some patterns that can help us figure out who survived and who didn't. This looks like a great machine learning problem!

<hr>
## Getting the data
Kaggle, a platform for data science competitions, has kindly compiled a dataset that is perfect for our needs and put it on their [website](https://www.kaggle.com/c/titanic/data) for budding machine learning enthusiasts to use. The training dataset tells us who survived and who didn't so we can use it to train our model. The test set doesn't tell us the fate of the passengers - that's what we're supposed to predict!

After downloading the datasets, we import the data and a few libraries that we will use later on.  

In [1]:
# for data analysis and wrangling
import pandas as pd
import numpy as np
from fancyimpute import KNN
from math import ceil

# for visualizations
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# for machine learning
#from sklearn.ensemble import RandomForestClassifier

# import data 
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test.csv")

WARNING (theano.configdefaults): g++ not available, if using conda: `conda install m2w64-toolchain`
C:\Users\Example\Anaconda3\lib\site-packages\theano\configdefaults.py:560: UserWarning: DeprecationWarning: there is no c++ compiler.This is deprecated and with Theano 0.11 a c++ compiler will be mandatory
  warnings.warn("DeprecationWarning: there is no c++ compiler."
WARNING (theano.configdefaults): g++ not detected ! Theano will be unable to execute optimized C-implementations (for both CPU and GPU) and will default to Python implementations. Performance will be severely degraded. To remove this warning, set Theano flags cxx to an empty string.
WARNING (theano.tensor.blas): Using NumPy C-API based implementation for BLAS functions.


<hr>
## Exploring the data

Now let's take a look at the data:

In [3]:
display(train_data.head())
print('_'*125)
display(test_data.head())

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

_____________________________________________________________________________________________________________________________


PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

#### What do **Pclass**, **SibSp**, and **Parch** mean? <br>
According to Kaggle, **Pclass** = passenger class, **Sibsp** = # of siblings/spouses aboard the Titanic, **Parch** = # of parents/children aboard the Titanic. 
<br>

#### What are the important feature types? 
- Categorical features: **Survived, Sex, Embarked, Pclass**
- Numerical features:
  - Discrete: **SibSp, Parch**
  - Continuous : **Fare, Age**
- Alphanumeric features: **Cabin, Ticket**


In [ ]:
# to find out size of data
print("training data dimensions:", train_data.shape)

In [ ]:
# What are the data types? Are there missing values?
train_data.info()
print('_'*125)
test_data.info()

#### What are the missing values? 
- From training set:
    - 687 missing **Cabin** values
    - 177 missing **Age** values
    - 2 missing **Embarked** values
- From test set:
    - 327 missing **Cabin** values
    - 86 missing **Age** values

In [ ]:
# Summarize integer and float type features
train_data.describe()
# Show more details about specific features
#data_train.describe(percentiles=[], include=["Pclass"])
#data_train.describe(percentiles=[], include=["Pclass"])
#data_train.describe(percentiles=[], include=["Pclass"])
#data_train.describe(percentiles=[], include=["Pclass"])
#data_train.describe(percentiles=[], include=["Pclass"])"""

#### What to numerical features tell us?
- The sample's survival rate is ~38% which is a bit higher than the actual 32%
- Most passengers on board were in 3rd class, while less that 25% where in 1st class
- More than 75% of passengers were less that 38 years old and the mean age was 30 
- More than 75% of passengers did not travel with their kids or their parents


In [ ]:
# Summarize object type features
train_data.describe(include=["O"])

### What this tells us

Hmm, it looks like we have some missing em ages and lots of missing cabin numbers. We will have to figure out ways to deal with those later on. 

<hr>
## Picking an algorithm

<hr>
## Preparing the data for machine learning algorithms
First we start by combining the training dataset and the test dataset so that we can edit them both together and ensure they end up in the same format. 


In [4]:
# Store the survival data
targets = train_data.Survived

# Drop survival data so training set and test set have the same shape and can be combined
train_data_dropped = train_data.drop(["Survived"], 1)

# Combine train and test data
combined_data = train_data_dropped.append(test_data)

combined_data.shape

(1309, 11)

The training data had 891 entries and the test data had 418 entries. $418 + 891 = 1309$ entires, so this looks good!

### Extracting passenger titles: 

If you look back at the data you will notice that passenger titles are always preceded by a comma and followed by a period. So we can create a function that splits the Name value at the comma and at the period to get the title.

In [5]:
# Extract titles from names and place them in a new column
combined_data["Title"] = combined_data["Name"].map(lambda name: name.split(",")[1].split(".")[0].strip())

# Drop names column because we no longer need it
#combined_data.drop(["Name"], 1, inplace=True)

# Show all different values in the Title column
combined_data.Title.value_counts()

Mr              757
Miss            260
Mrs             197
Master           61
Dr                8
Rev               8
Col               4
Ms                2
Mlle              2
Major             2
Dona              1
Lady              1
the Countess      1
Capt              1
Jonkheer          1
Mme               1
Sir               1
Don               1
Name: Title, dtype: int64

In [ ]:
#back up copy
back_up2 = combined_data.copy()

Looks like it worked!
<br>
We can also group similar titles together to simplify our model and reduce the risk of "overfitting." Overfitting happens when a model is too complex in such a way that it is so good at predicting training data outcomes but not so good at predicting outcomes for unseen data. For example, it would be nice if our model knows how to predict outcomes for countesses, but chances are, there aren't that many countesses on the Titanic, so it would be better to group Royalties  together so that our model generalizes better when looking at new data.   

In [6]:
Title_dict = {
    "Mr":"Mr",
    "Miss":"Ms",
    "Mrs":"Mrs",
    "Master":"Master",
    "Dr":"Other",
    "Rev":"Other",
    "Col":"Military",
    "Ms":"Ms",
    "Mlle":"Ms",
    "Major":"Military",
    "Don":"Royal or Noble",
    "the Countess":"Royal or Noble",
    "Lady":"Royal or Noble",
    "Dona":"Royal or Noble",
    "Sir":"Royal or Noble",
    "Mme":"Mrs",
    "Jonkheer":"Royal or Noble",
    "Capt":"Other"
}

# Group titles using mappings in dictionary above
combined_data["Title"] = combined_data["Title"].map(Title_dict) 

combined_data["Title"].value_counts()

Mr                757
Ms                264
Mrs               198
Master             61
Other              17
Military            6
Royal or Noble      6
Name: Title, dtype: int64

### Processing passenger ages:

If you recall from the data exploration step, there were about 177 and 86 missing age values from the training set and the data set respectively. We know that age is an important factor in determining survival so we need to come up with a way to fill in the missing ages.

Here we are going to group people by their gender, class, and title and then use these groupings to determine the missing values.

In [7]:
# Select train data and group by Sex, class, and title in that order
grouped_train_data = combined_data.head(891).groupby(["Sex","Pclass", "Title"])

# Select test data and group by Sex, class, and title in that order
grouped_test_data = combined_data.iloc[891:].groupby(["Sex", "Pclass", "Title"])

# Find and display medians 
display(grouped_train_data.median())
print('_'*125)
display(grouped_test_data.median())

PassengerId   Age  SibSp  Parch       Fare
Sex    Pclass Title                                                     
female 1      Mrs                   499.0  40.0    1.0    0.0   79.20000
              Ms                    369.0  30.0    0.0    0.0   88.25000
              Other                 797.0  49.0    0.0    0.0   25.92920
              Royal or Noble        658.5  40.5    0.5    0.0   63.05000
       2      Mrs                   438.0  32.0    1.0    0.0   26.00000
              Ms                    444.0  24.0    0.0    0.0   13.00000
       3      Mrs                   405.5  31.0    1.0    1.0   15.97500
              Ms                    372.0  18.0    0.0    0.0    8.75625
male   1      Master                446.0   4.0    1.0    2.0  120.00000
              Military              592.5  54.0    0.0    0.0   28.52500
              Mr                    463.0  40.0    0.0    0.0   42.40000
              Other                 661.0  47.0    1.0    0.0   71.00000
              Royal or Noble        600.0  40.0    0.0    0.0   27.72080
       2      Master                408.0   1.0    1.0    1.0   26.00000
              Mr                    440.0  31.0    0.0    0.0   13.00000
              Other                 358.5  46.5    0.0    0.0   13.00000
       3      Master                270.5   4.0    3.5    1.0   28.51250
              Mr                    472.0  26.0    0.0    0.0    7.89580

_____________________________________________________________________________________________________________________________


PassengerId   Age  SibSp  Parch       Fare
Sex    Pclass Title                                                     
female 1      Mrs                  1076.0  48.0    1.0    0.0   63.35830
              Ms                   1074.0  32.0    0.0    0.0  158.20835
              Royal or Noble       1306.0  39.0    0.0    0.0  108.90000
       2      Mrs                  1123.5  29.0    0.0    0.0   26.00000
              Ms                   1121.0  19.5    1.0    1.0   24.50000
       3      Mrs                  1051.0  28.0    1.0    1.0   14.45420
              Ms                   1089.0  22.0    0.0    0.0    7.87920
male   1      Master               1022.0   9.5    1.0    2.0  198.43750
              Military             1058.5  50.0    0.5    0.0  128.01250
              Mr                   1102.0  42.0    0.0    0.0   50.24790
              Other                1185.0  53.0    1.0    1.0   81.85830
       2      Master               1033.5   5.0    0.5    1.5   27.75000
              Mr                   1156.0  28.0    0.0    0.0   13.00000
              Other                1048.5  35.5    0.5    0.5   19.50000
       3      Master               1173.0   7.0    1.0    1.0   15.24580
              Mr                   1101.0  25.0    0.0    0.0    7.85420

So the function we want to create to fill in the missing ages first checks the passenger's age, then their class, then their title and uses that info to determine what age to give them. If we haven't seen that title before, we should just plug in the median age.

 <div class="alert alert-block alert-warning">Note that we have to be super careful not to introduce any information from the test data into the training data. The point of a predictive machine learning model is to make accurate predictions about new data that is *unseen* during the training.</div>

In [ ]:
# Function to round approximations of missing ages to nearest 0.5             
def round_age(age):
    return round(age * 2) / 2


# Function that fills in missing ages
def age_filler(incomplete_data, grouped_data, Training=True, size=0):
    # Change index to PassengerId so that each passenger has a unique index
    if Training != True:
        incomplete_data = incomplete_data.reset_index()   
    
    for row in grouped_data:
        try:
            # This is a data frame the has the age and fares of passengers sharing the same gender, class, and title
            subset_incomplete_data = incomplete_data.loc[(incomplete_data["Sex"]==row[0][0]) 
                                                        & (incomplete_data["Pclass"]==row[0][1]) 
                                                        & (incomplete_data["Title"]==row[0][2]), 
                                                        (["Age", "Fare"])]
        # Skip to next row if no values are found
        except KeyError:
            continue
            
        # Array of global indexes of passengers in subset with missing ages
        missing_ages_global_index = subset_incomplete_data.index[subset_incomplete_data["Age"].isnull()].tolist()
        missing_ages_global_index.sort()

        # Array of local indexes of passengers in subset with missing ages
        missing_ages_local_index = np.where(subset_incomplete_data["Age"].isnull())[0]
        missing_ages_local_index.sort()

        # Get number of passengers in subset_train_data, use this number to calculate number of KNN neighbours 
        subset_incomplete_data_size = subset_incomplete_data.shape[0]
        
        try:
            # Use KNN to fill in missing ages based on similarities between passegers' fares
            # Returns an unindexed but ordered numpy array of ages
            subset_complete_data = KNN(k=ceil(subset_incomplete_data_size*0.05)).complete(subset_incomplete_data)
        # Handles case when there are no missing values
        except ValueError:
            continue
            
        counter = 0
        # Iterate over estimated ages in subset and place them back in the original dataset
        # Use indexes to match values from subset to original data set
        for passenger_local_index in missing_ages_local_index:
            passenger_estimated_age = round_age(subset_complete_data[passenger_local_index][0])
            passenger_global_index = missing_ages_global_index[counter]
            counter+=1
            completed_data = incomplete_data.set_value(passenger_global_index, "Age", passenger_estimated_age)
    # Handles test set
    if Training != True:
        return completed_data.iloc[size:] 
    # Handles training set
    else:
        return completed_data

train_data = combined_data.head(891)

# Estimate missing ages in training set, pass in a copy so that original data set is not changed
train_data_filled = age_filler(train_data.copy(), grouped_train_data)

# Estimate missing ages in test set
test_data_filled = age_filler(combined_data.copy(), grouped_train_data, Training = False, size = 891)

# Combining training and test set after filling in all ages
combined_data = train_data_filled.append(test_data_filled).reset_index()

In [9]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
level_0        1309 non-null int64
Age            1309 non-null float64
Cabin          295 non-null object
Embarked       1307 non-null object
Fare           1308 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Ticket         1309 non-null object
Title          1309 non-null object
index          418 non-null float64
dtypes: float64(3), int64(5), object(6)
memory usage: 143.2+ KB


### Processing Fare and Embarked features

In this data set we have one missing Fare value and two missing Embarked values, so let's just fill them in directly.
<br><br>
First: let's look at the passenger with a missing fare

In [10]:
missing_fare_index = combined_data.index[combined_data["Fare"].isnull()]
combined_data.iloc[missing_fare_index]

level_0   Age Cabin Embarked  Fare                Name  Parch  \
1043     1043  60.5   NaN        S   NaN  Storey, Mr. Thomas      0   

      PassengerId  Pclass   Sex  SibSp Ticket Title  index  
1043         1044       3  male      0   3701    Mr  152.0

This passenger is a male in 3rd class whose title is Mr.
Let's find the subset of passengers who also have the same gender, class, and title as this passenger and then assign him the median fare of that subset.

In [11]:
subset = combined_data.loc[(combined_data["Sex"]=="male")
                          & (combined_data["Pclass"]==3)
                          & (combined_data["Title"]=="Mr")]

# Fill empty fare with median
combined_data["Fare"].fillna(float(subset["Fare"].median()), inplace=True)

In [12]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
level_0        1309 non-null int64
Age            1309 non-null float64
Cabin          295 non-null object
Embarked       1307 non-null object
Fare           1309 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Ticket         1309 non-null object
Title          1309 non-null object
index          418 non-null float64
dtypes: float64(3), int64(5), object(6)
memory usage: 143.2+ KB


Now we do a similar thing for the missing Embarked values, but we replace missing values with the mode instead.
We also make sure we do not leak data from the test set to the training set.

In [13]:
missing_embarked_index = combined_data.index[combined_data["Embarked"].isnull()]
missing_embarked_rows = combined_data.iloc[missing_embarked_index]
display(missing_embarked_rows)

level_0   Age Cabin Embarked  Fare  \
61        61  38.0   B28      NaN  80.0   
829      829  62.0   B28      NaN  80.0   

                                          Name  Parch  PassengerId  Pclass  \
61                         Icard, Miss. Amelie      0           62       1   
829  Stone, Mrs. George Nelson (Martha Evelyn)      0          830       1   

        Sex  SibSp  Ticket Title  index  
61   female      0  113572    Ms    NaN  
829  female      0  113572   Mrs    NaN

In [14]:
for index, row in missing_embarked_rows.iterrows():
    subset = combined_data.head(891).loc[(combined_data.head(891)["Sex"]==row["Sex"])
                                        & (combined_data.head(891)["Pclass"]==row["Pclass"])
                                        & (combined_data.head(891)["Title"]==row["Title"])]                                    
    # Fill empty fields with mode
    combined_data.set_value(index, "Embarked", str(subset["Embarked"].mode().iloc[0])) 

In [15]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
level_0        1309 non-null int64
Age            1309 non-null float64
Cabin          295 non-null object
Embarked       1309 non-null object
Fare           1309 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Ticket         1309 non-null object
Title          1309 non-null object
index          418 non-null float64
dtypes: float64(3), int64(5), object(6)
memory usage: 143.2+ KB


Fare and Emarked now have no missing values.

### Creating Family_Size

Now we are going to create the **FamilySize** feature by adding **SibSp** and **Parch** and 1. This is makes sense because a **FamilySize** feature might be a better predictor than either of the other features separatley. Families tend to stick together, which probably affected their chanced of survival. We then drop the **SibSp** and **Parch** columns because the information the contain are summarized by the new **FamilySize** feature.

In [16]:
combined_data["Family_Size"] = combined_data["SibSp"] + combined_data["Parch"] + 1

# Drop all columns we don't need
combined_data_dropped = combined_data.drop(["Name", "SibSp", "Parch", "Cabin", "Ticket", "PassengerId", "index","level_0"], 1)

train_data_final = combined_data_dropped.head(891).copy().reset_index().drop(["index"],1)
display(train_data_final.head())
test_data_final = combined_data_dropped.iloc[891:].copy()
display(test_data_final.head())

Age Embarked     Fare  Pclass     Sex Title  Family_Size
0  22.0        S   7.2500       3    male    Mr            2
1  38.0        C  71.2833       1  female   Mrs            2
2  26.0        S   7.9250       3  female    Ms            1
3  35.0        S  53.1000       1  female   Mrs            2
4  35.0        S   8.0500       3    male    Mr            1

Age Embarked     Fare  Pclass     Sex Title  Family_Size
891  34.5        Q   7.8292       3    male    Mr            1
892  47.0        S   7.0000       3  female   Mrs            2
893  62.0        Q   9.6875       2    male    Mr            1
894  27.0        S   8.6625       3    male    Mr            1
895  22.0        S  12.2875       3  female   Mrs            3

## ML

In [19]:
import h2o
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  ------------------------------
H2O cluster uptime:         17 secs
H2O cluster version:        3.10.4.8
H2O cluster version age:    6 months and 12 days !!!
H2O cluster name:           H2O_from_python_Example_am1iva
H2O cluster total nodes:    1
H2O cluster free memory:    237.2 Mb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
Python version:             3.6.2 final
--------------------------  ------------------------------

We convert the data we are working with into H2O dataframes so that the H2O algorithms can work with them. We also append the survival data back to the training set.

In [22]:
# We append the survival data back to the training set
train_data_final_surv = train_data_final.join(targets) 

#train_data_final_surv["Survived"] = train_data_final_surv["Survived"].astype("category")

train_data_final_surv.head()

Age Embarked     Fare  Pclass     Sex Title  Family_Size Survived
0  22.0        S   7.2500       3    male    Mr            2        0
1  38.0        C  71.2833       1  female   Mrs            2        1
2  26.0        S   7.9250       3  female    Ms            1        1
3  35.0        S  53.1000       1  female   Mrs            2        1
4  35.0        S   8.0500       3    male    Mr            1        0

In [23]:
# Convert to H2O dataframes
h2o_train_data = h2o.H2OFrame(train_data_final_surv)
h2o_test_data = h2o.H2OFrame(test_data_final)

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


In [29]:
# We make sure the survival values (1 and 0) are treated as categorical variables
h2o_train_data["Survived"] = h2o_train_data["Survived"].asfactor() 
h2o_train_data.types

{'Age': 'real',
 'Embarked': 'enum',
 'Family_Size': 'int',
 'Fare': 'real',
 'Pclass': 'int',
 'Sex': 'enum',
 'Survived': 'enum',
 'Title': 'enum'}

In [37]:
# Save data frames so we can load them again without going through whole analysis
#h2o.download_csv(h2o_train_data,"h2o_train_data.csv")
#h2o.download_csv(h2o_test_data,"h2o_test_data.csv")
h2o_train_data = h2o.import_file("h2o_train_data.csv")
h2o_test_data = h2o.import_file("h2o_test_data.csv")

In [32]:
# Get names of columns that contain the features we want to train the algorithm on
feature_names = h2o_train_data.col_names[0:-1]
# Get name of column that has targets
targets_name = h2o_train_data.col_names[-1]

Survived


In [33]:
from h2o.estimators.random_forest import H2ORandomForestEstimator

In [84]:
# Now let's create a random forest instance
RF = H2ORandomForestEstimator(ntrees=400,
                              nfolds=10,
                              binomial_double_trees=True,
                              stopping_metric = "AUC",
                              stopping_tolerance = 0,
                              stopping_rounds = 15)
                              
                              #stopping_tolerance = 0.001, #0.1% after 9 trees
                              #stopping_rounds = 3,
                              #score_tree_interval = 3)
#model_id ="RF_200t_6nf_s2", ntrees=200, nfolds=6, seed=2
# binomial_double_trees=True,
#stopping_metric="misclassification",
#stopping_rounds=3,
#stopping_tolerance=0.02,
#------
#sample_rate=0-1
#col_sample_rate_per_tree=0-1

Here nfolds refers the the number of folds we use for cross-validation. This basically allows us to train multiple models using different subset of the data to get a better sense of the model's accuracy.
<br>

There are tens of extra parameters we can adjust here. We can choose the number of trees, the depth of each tree, the number of variables to decide each split, and so on. There is a very useful class called GridSearch that trains different models with different parameters and allows us to pick the optimal parameters. 

In [ ]:
# First we define the parameters we want to try
hyper_params={"max_depth": [20, 40, 60],
              "min_rows": [1, 2, 3, 4],
              "sample_rate": [0.5, 0.6, 0.7, 0.8],
              "mtries": [1, 2, 3, 4],
              "col_sample_rate_per_tree": [0.8, 0.9, 1]}

# Then we create a grid_search instance
grid_search = h2o.grid.H2OGridSearch(RF, hyper_parameters)


In [85]:
RF.train(feature_names, targets_name, training_frame=h2o_train_data)

drf Model Build progress: |███████████ (cancelled)


H2OJobCancelled: Job<$03017f00000132d4ffffffff$_b78fab6bc5bc89d2167cad800b974345> was cancelled by the user.

In [75]:
RF
#.model_performance(h2o_train_data)

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  DRF_model_python_1512328784826_5994


ModelMetricsBinomial: drf
** Reported on train data. **

MSE: 0.127435504337922
RMSE: 0.35698109801209643
LogLoss: 0.4221182466499098
Mean Per-Class Error: 0.18353944971718916
AUC: 0.8755179539620149
Gini: 0.7510359079240299
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.6042296603436913: 


0    1    Error    Rate
-----  ---  ---  -------  -------------
0      508  41   0.0747   (41.0/549.0)
1      100  242  0.2924   (100.0/342.0)
Total  608  283  0.1582   (141.0/891.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.60423      0.7744    149
max f2                       0.179571     0.810811  272
max f0point5                 0.767001     0.827896  106
max accuracy                 0.60423      0.841751  149
max precision                0.999899     1         0
max recall                   0.00945347   1         398
max specificity              0.999899     1         0
max absolute_mcc             0.60423      0.661156  149
max min_per_class_accuracy   0.335065     0.804094  218
max mean_per_class_accuracy  0.60423      0.816461  149

Gains/Lift Table: Avg response rate: 38.38 %



group    cumulative_data_fraction    lower_threshold    lift       cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ---------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0123457                   0.998482           2.60526    2.60526            1                1                           0.0321637       0.0321637                  160.526   160.526
    2        0.0213244                   0.998099           2.60526    2.60526            1                1                           0.0233918       0.0555556                  160.526   160.526
    3        0.030303                    0.997043           2.60526    2.60526            1                1                           0.0233918       0.0789474                  160.526   160.526
    4        0.040404                    0.996034           2.60526    2.60526            1                1                           0.0263158       0.105263                   160.526   160.526
    5        0.0516274                   0.994717           2.60526    2.60526            1                1                           0.0292398       0.134503                   160.526   160.526
    6        0.10101                     0.98092            2.60526    2.60526            1                1                           0.128655        0.263158                   160.526   160.526
    7        0.150393                    0.954314           2.60526    2.60526            1                1                           0.128655        0.391813                   160.526   160.526
    8        0.200898                    0.908441           2.60526    2.60526            1                1                           0.131579        0.523392                   160.526   160.526
    9        0.300786                    0.710072           2.57599    2.59554            0.988764         0.996269                    0.25731         0.780702                   157.599   159.554
    10       0.401796                    0.371916           1.91053    2.42333            0.733333         0.930168                    0.192982        0.973684                   91.0526   142.333
    11       0.501684                    0.18601            0.234181   1.98746            0.0898876        0.762864                    0.0233918       0.997076                   -76.5819  98.746
    12       0.600449                    0.0914339          0.0296053  1.66542            0.0113636        0.639252                    0.00292398      1                          -97.0395  66.5421
    13       0.700337                    0.0540869          0          1.42788            0                0.548077                    0               1                          -100      42.7885
    14       0.800224                    0.0345201          0          1.24965            0                0.479663                    0               1                          -100      24.9649
    15       0.900112                    0.0199041          0          1.11097            0                0.426434                    0               1                          -100      11.0973
    16       1                           0.00662293         0          1                  0                0.383838                    0               1                          -100      0



ModelMetricsBinomial: drf
** Reported on cross-validation data. **

MSE: 0.13093557001117437
RMSE: 0.3618502038291182
LogLoss: 0.4399389772996845
Mean Per-Class Error: 0.18756590930879113
AUC: 0.8690788142183024
Gini: 0.7381576284366047
Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.6109921179219233: 


0    1    Error    Rate
-----  ---  ---  -------  -------------
0      510  39   0.071    (39.0/549.0)
1      104  238  0.3041   (104.0/342.0)
Total  614  277  0.1605   (143.0/891.0)

Maximum Metrics: Maximum metrics at their respective thresholds



metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.610992     0.768982  147
max f2                       0.173717     0.809498  274
max f0point5                 0.636367     0.822695  140
max accuracy                 0.61862      0.839506  145
max precision                1            1         0
max recall                   0.00920208   1         396
max specificity              1            1         0
max absolute_mcc             0.61862      0.656695  145
max min_per_class_accuracy   0.336876     0.798246  216
max mean_per_class_accuracy  0.610992     0.812434  147

Gains/Lift Table: Avg response rate: 38.38 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.010101                    0.998166           2.60526   2.60526            1                1                           0.0263158       0.0263158                  160.526   160.526
    2        0.020202                    0.997622           2.02632   2.31579            0.777778         0.888889                    0.0204678       0.0467836                  102.632   131.579
    3        0.030303                    0.99629            2.60526   2.41228            1                0.925926                    0.0263158       0.0730994                  160.526   141.228
    4        0.040404                    0.994872           2.60526   2.46053            1                0.944444                    0.0263158       0.0994152                  160.526   146.053
    5        0.0505051                   0.992693           2.60526   2.48947            1                0.955556                    0.0263158       0.125731                   160.526   148.947
    6        0.10101                     0.966698           2.43158   2.46053            0.933333         0.944444                    0.122807        0.248538                   143.158   146.053
    7        0.150393                    0.925072           2.36842   2.43028            0.909091         0.932836                    0.116959        0.365497                   136.842   143.028
    8        0.200898                    0.859748           2.43158   2.43061            0.933333         0.932961                    0.122807        0.488304                   143.158   143.061
    9        0.300786                    0.632163           1.93199   2.26502            0.741573         0.869403                    0.192982        0.681287                   93.1993   126.502
    10       0.400673                    0.41003            0.907451  1.92658            0.348315         0.739496                    0.0906433       0.77193                    -9.25488  92.6581
    11       0.500561                    0.234814           0.731815  1.68816            0.280899         0.647982                    0.0730994       0.845029                   -26.8185  68.8164
    12       0.600449                    0.138702           0.526907  1.49498            0.202247         0.573832                    0.0526316       0.897661                   -47.3093  49.4983
    13       0.700337                    0.0819259          0.292726  1.32351            0.11236          0.508013                    0.0292398       0.926901                   -70.7274  32.3507
    14       0.800224                    0.047279           0.175636  1.18022            0.0674157        0.453015                    0.0175439       0.944444                   -82.4364  18.0224
    15       0.901235                    0.024131           0.231579  1.0739             0.0888889        0.412204                    0.0233918       0.967836                   -76.8421  7.39005
    16       1                           0                  0.325658  1                  0.125            0.383838                    0.0321637       1                          -67.4342  0


Cross-Validation Metrics Summary: 


mean      sd         cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid    cv_6_valid    cv_7_valid    cv_8_valid
-----------------------  --------  ---------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
accuracy                 0.85437   0.0220131  0.855856      0.778689      0.872549      0.877359      0.850467      0.886792      0.863248      0.85
auc                      0.86451   0.0337858  0.778283      0.798951      0.865283      0.904875      0.880224      0.909654      0.863175      0.915637
err                      0.14563   0.0220131  0.144144      0.221311      0.127451      0.122642      0.149533      0.113208      0.136752      0.15
err_count                16.375    3.14121    16            27            13            13            16            12            16            18
f0point5                 0.822809  0.0295986  0.791667      0.763547      0.868263      0.856482      0.793269      0.875576      0.852941      0.780731
f1                       0.794924  0.0421868  0.703704      0.696629      0.816901      0.850575      0.804878      0.863636      0.783784      0.839286
f2                       0.77388   0.066485   0.633333      0.640496      0.771277      0.844749      0.816832      0.852018      0.725         0.907336
lift_top_group           2.64383   0.242631   3.46875       2.39216       2.61538       2.40909       2.675         2.35556       2.78571       2.44898
logloss                  0.438423  0.0555803  0.487298      0.608687      0.479473      0.378519      0.409138      0.373667      0.417551      0.353054
max_per_class_error      0.259917  0.0665379  0.40625       0.392157      0.25641       0.159091      0.175         0.155556      0.309524      0.225352
mcc                      0.690462  0.0481698  0.631512      0.542394      0.728903      0.74674       0.684318      0.767527      0.700007      0.722297
mean_per_class_accuracy  0.833912  0.0302218  0.777888      0.754626      0.847985      0.872067      0.845336      0.881239      0.825238      0.866916
mean_per_class_error     0.166088  0.0302218  0.222112      0.245374      0.152015      0.127933      0.154664      0.118761      0.174762      0.133084
mse                      0.130223  0.0172465  0.149291      0.184823      0.126938      0.109753      0.126584      0.109022      0.127876      0.107502
precision                0.845982  0.0385227  0.863636      0.815789      0.90625       0.860465      0.785714      0.883721      0.90625       0.746032
r2                       0.441928  0.0816502  0.272385      0.240291      0.462491      0.547953      0.45923       0.553746      0.444289      0.555037
recall                   0.76315   0.0841838  0.59375       0.607843      0.74359       0.840909      0.825         0.844444      0.690476      0.959184
rmse                     0.359413  0.0228628  0.386381      0.42991       0.356283      0.33129       0.355787      0.330184      0.357597      0.327875
specificity              0.904674  0.0412654  0.962025      0.901408      0.952381      0.903226      0.865672      0.918033      0.96          0.774648

Scoring History: 


timestamp            duration          number_of_trees    training_rmse        training_logloss    training_auc        training_lift       training_classification_error
---  -------------------  ----------------  -----------------  -------------------  ------------------  ------------------  ------------------  -------------------------------
     2017-12-03 17:21:01  57.818 sec        0.0                nan                  nan                 nan                 nan                 nan
     2017-12-03 17:21:01  57.861 sec        1.0                0.42359906474772235  5.089755026975596   0.812305900621118   2.4691673212882956  0.1724137931034483
     2017-12-03 17:21:01  57.909 sec        2.0                0.4212999476155483   4.942451182732159   0.811720601072453   2.5813616610333177  0.18582375478927204
     2017-12-03 17:21:01  57.957 sec        3.0                0.417450103706241    4.469764368027364   0.810405643738977   2.605263157894737   0.18597560975609756
     2017-12-03 17:21:01  58.004 sec        4.0                0.4166070613488154   4.276242830608549   0.8136200716845878  2.605263157894737   0.18983957219251338
---  ---                  ---               ---                ---                  ---                 ---                 ---                 ---
     2017-12-03 17:21:04  1 min  1.171 sec  34.0               0.36344688924662977  0.6808998433279547  0.8719841498098616  2.605263157894737   0.17059483726150393
     2017-12-03 17:21:04  1 min  1.347 sec  35.0               0.36391971852979665  0.6486683152762677  0.8713743222658954  2.605263157894737   0.1739618406285073
     2017-12-03 17:21:04  1 min  1.518 sec  36.0               0.3633721763312615   0.6483882905164537  0.8712331831399994  2.605263157894737   0.1739618406285073
     2017-12-03 17:21:05  1 min  1.707 sec  37.0               0.36346214967973184  0.615276426577736   0.8722717540664047  2.605263157894737   0.1717171717171717
     2017-12-03 17:21:08  1 min  5.011 sec  259.0              0.35698109801209643  0.4221182466499098  0.8755179539620149  2.605263157894737   0.15824915824915825


See the whole table with table.as_data_frame()
Variable Importances: 


variable     relative_importance    scaled_importance    percentage
-----------  ---------------------  -------------------  ------------
Title        16446.1                1                    0.244488
Fare         12855.4                0.781672             0.191109
Age          12655.6                0.769519             0.188138
Sex          11858.8                0.721071             0.176293
Pclass       6632.97                0.403316             0.098606
Family_Size  4902.59                0.298101             0.0728821
Embarked     1916.02                0.116503             0.0284836

Now we finally make our predictions on the test set and prepare it for submission

In [76]:
predictions = RF.predict(h2o_test_data)

drf prediction progress: |████████████████████████████████████████████████| 100%


In [39]:
predictions.head

predict p0 p1 
 
 
 0 0.878779 0.121221 
 0 0.705082 0.294918 
 0 0.706405 0.293595 
 0 0.467629 0.532371 
 1 0.216161 0.783839 
 0 0.929172 0.070828 
 0 0.567109 0.432891 
 0 0.955731 0.0442692 
 1 0.15317 0.84683 
 0 0.947986 0.0520137

<bound method H2OFrame.head of >

In [77]:
# Convert H2O dataframe back to pandas dataframe
predictions = predictions.as_data_frame()

predictions.drop(["p0","p1"], 1, inplace=True)

# Create a PassengerId column
predictions.insert(0, "PassengerId", range(892, 1310))

# Rename predict column to Survived 
predictions.rename(columns={"predict": "Survived"}, inplace=True)

predictions.head()

PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1

In [78]:
# Export predictions to csv file
predictions.to_csv("predictions.csv", index=False)

In [56]:
h2o.cluster().shutdown()

[WARNING] in <ipython-input-56-bd1e984f485f> line 1:
    >>> h2o.shutdown()
        ^^^^ Deprecated, use ``h2o.cluster().shutdown()``.
H2O session _sid_ad35 closed.
